vpn/src/main/java/com/duckduckgo/mobile/android/vpn/prefs/VpnSharedPreferencesProvider.kt

```diff
@@ -19,24 +19,75 @@ package com.duckduckgo.mobile.android.vpn.prefs
 import com.squareup.anvil.annotations.ContributesBinding
+import timber.log.Timber
 import javax.inject.Inject
 
 interface VpnSharedPreferencesProvider {
-    fun getSharedPreferences(name: String, multiprocess: Boolean = false): SharedPreferences
+    /**
+     * Returns an instance of Shared preferences
+     * @param name Name of the shared preferences
+     * @param multiprocess `true` if the shared preferences will be accessed from several processes else `false`
+     * @param migrate `true` if the shared preferences existed prior to use the [VpnSharedPreferencesProvider], else `false`
+     */
+    fun getSharedPreferences(name: String, multiprocess: Boolean = false, migrate: Boolean = false): SharedPreferences
 }
 
+private const val MIGRATED_TO_HARMONY = "migrated_to_harmony"
+
 @ContributesBinding(AppScope::class)
 class VpnSharedPreferencesProviderImpl @Inject constructor(
     private val context: Context
 ) : VpnSharedPreferencesProvider {
-    override fun getSharedPreferences(name: String, multiprocess: Boolean): SharedPreferences {
+    override fun getSharedPreferences(name: String, multiprocess: Boolean, migrate: Boolean): SharedPreferences {
         return if (multiprocess) {
-            context.getHarmonySharedPreferences(name)
+            if (migrate) {
+                Timber.v("Migrate and return preferences to Harmony")
+                return migrateToHarmonyIfNecessary(name)
+            } else {
+                Timber.v("Return Harmony preferences")
+                context.getHarmonySharedPreferences(name)
+            }
         } else {
             context.getSharedPreferences(name, MODE_PRIVATE)
         }
     }
+
+    private fun migrateToHarmonyIfNecessary(name: String): SharedPreferences {
+        val origin = context.getSharedPreferences(name, MODE_PRIVATE)
+        val destination = context.getHarmonySharedPreferences(name)
+
+        if (destination.getBoolean(MIGRATED_TO_HARMONY, false)) return destination
+        Timber.v("Performing migration to Harmony")
+
+        val contents = origin.all
+
+        contents.keys.forEach { key ->
+            when (val originalValue = contents[key]) {
+                is Boolean -> {
+                    destination.edit { putBoolean(key, originalValue) }
+                }
+                is Long -> {
+                    destination.edit { putLong(key, originalValue) }
+                }
+                is Int -> {
+                    destination.edit { putInt(key, originalValue) }
+                }
+                is Float -> {
+                    destination.edit { putFloat(key, originalValue) }
+                }
+                is String -> {
+                    destination.edit { putString(key, originalValue) }
+                }
+                else -> Timber.w("Could not migrate $key from $name preferences")
+            }
+        }
+
+        destination.edit(commit = true) { putBoolean(MIGRATED_TO_HARMONY, true) }
```

Comment:
Thanks for the feedback @pablobaxter. Harmony came in the right moment for us. We were feeling a bit of pain because how shared prefs deal with multiprocess. The library is great.

> I would recommend storing both preferences in a class variable, and returning the correct object via this getSharedPreferences()

Do you mean having like a `Map<String, SharedPreference>` to have those shared prefs instances in memory so that we don't need to go through `context.getSharedPreferences` or `context.getHarmonySharedPreferences` every time?

> all or nothing migration

huh, that's a good point. A tricky one also. The main issue for us is that we know shared preferences doesn't deal with multi-process properly, causing bugs because many times values are not synced across processes when we need it.
1. If we do all or nothing migration, when it's nothing we'll returned the android shared prefs, which we know it will fail sometimes.
2. If we don't do all or nothing and (right now) assumed migration will just be OK, it may not be. The result in this case is that the first value may not be correct, but the subsequent ones would be, given than Harmony is much more reliable than shared prefs for multi-process.

I think I prefer 2/ over 1/. Does that make sense?

Reply:
> Do you mean having like a Map<String, SharedPreference> to have those shared prefs instances in memory so that we don't need to go through context.getSharedPreferences or context.getHarmonySharedPreferences every time?

Ah, never mind. You are calling for the preference file via the function you provided. Both Harmony and Android already store the prefs implementation as a singleton internally. What I was alluding to was that if you call for the prefs first, then immediately call any `get` function, it could block until the underlying prefs file has loaded. This is true for both SharedPreferences and Harmony. By eagerly calling for preferences, you give both Android and Harmony ample time to load preferences, so you don't risk a thread blocking `get` or `edit` call.

Basically, something like this would be bad:

```
fun getSomeValue(): String {
    val prefs = context.getSharedPreferences(PREFS_NAME, MODE_PRIVATE) // First call for prefs, which loads the prefs file under the covers
    return prefs.getString("blah", null) // This would most likely block if the prefs object didn't finish loading the file
}
```

But this would improve the performance of `getSomeValue()` in the example above:

```
private val prefs = context.getSharedPreferences(PREFS_NAME, MODE_PRIVATE) // First call for prefs, which loads the prefs file under the covers
fun getSomeValue(): String {
    return prefs.getString("blah", null) // Probably won't block if ample time occurred between the prefs call above and this call
}
```

> huh, that's a good point. A tricky one also. The main issue for us is that we know shared preferences doesn't deal with multi-process properly, causing bugs because many times values are not synced across processes when we need it.
>
>    If we do all or nothing migration, when it's nothing we'll returned the android shared prefs, which we know it will fail sometimes.
>    If we don't do all or nothing and (right now) assumed migration will just be OK, it may not be. The result in this case is that the first value may not be correct, but the subsequent ones would be, given than Harmony is much more reliable than shared prefs for multi-process.
>
> I think I prefer 2/ over 1/. Does that make sense?

The problem with 2 is that you might set `MIGRATED_TO_HARMONY` to `true` even though one of the values above may have failed (apply doesn't return a boolean status, and fails silently), in which case, you won't retry to migrate again. I'm not saying failures will happen (they are rare for both Harmony and SharedPreferences), but they can happen. It's a risk, but just pointing out the options here.